### PR TITLE
localBinInPath: Initial commit

### PR DESCRIPTION
**Pull request by Kreyren**
###### Motivation for this change
being annoyed with `bin` in my dir while missing FHS3_0

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

First commit and too confused by this checklist aAAaaAaAAAAA

### CODE REVIEW

File: nixos/modules/config/shells-environment.nix [hunk 1]
```diff
@@ -198,6 +206,11 @@ in
           # ~/bin if it exists overrides other bin directories.
           export PATH="$HOME/bin:$PATH"
         ''}
+
+        ${optionalString cfg.localBinInPath ''
+          # ~/.local/bin if it exists overrides other bin directories.
+          export PATH="$HOME/.local/bin:$PATH"
```
Comment (tengkuizdihar) [inline nixos/modules/config/shells-environment.nix:212]: Should the `L` be capital in `.local`?

Comment (Kreyren) [reply nixos/modules/config/shells-environment.nix:212]: why would it be capital.. FHS3_0 defines that as `.local` and nixOS seems to be using it?

image

Comment (tengkuizdihar) [reply nixos/modules/config/shells-environment.nix:212]: my bad, it seems i confuse the L font in github with the l font.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128271` at 197821f4 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.trivial</li>
  </ul>
</details>

**Comment by Mindavi** [comment]
I don't fully understand this.

- Why are the release notes for version 20.03 changed? (Note that version is yy.mm)
- Why do the release notes say that it's a change from before, while it seems like 'just' an addition?
- Why would you want ~/.local/bin in your PATH? That's not very pure.

**Comment by Kreyren** [comment]
> Why are the release notes for version 20.03 changed? (Note that version is yy.mm) @Mindavi

Because i didn't know that it's 20.03 x.x

> Why do the release notes say that it's a change from before, while it seems like 'just' an addition? @Mindavi

Wasn't sure if that's the way the release notes are done so i just copy-pasted it and changed the paths

> Why would you want ~/.local/bin in your PATH? That's not very pure. @Mindavi

Because `$HOME/bin` is distracting

**Comment by Mindavi** [comment]
No problem, now you know.

I personally think this is something that's better left in your own config. Don't think this is something that a lot of people require or want. But maybe someone else thinks otherwise.

**Comment by teto** [comment]
Likewise I dont think it's worth adding to nixpkgs, it's simple enough to prepend that to PATH.

File: nixos/modules/config/shells-environment.nix [hunk 2]
```diff
@@ -126,6 +126,14 @@ in
       type = types.bool;
     };
 
+    environment.localBinInPath = mkOption {
+      description = ''
+        Include ~/.local/bin/ in $PATH.
```
Comment (SuperSandro2000) [inline nixos/modules/config/shells-environment.nix:131]:
```suggestion
        Add ~/.local/bin/ to $PATH
```

File: nixos/modules/config/shells-environment.nix [hunk 3]
```diff
@@ -198,6 +206,11 @@ in
           # ~/bin if it exists overrides other bin directories.
           export PATH="$HOME/bin:$PATH"
         ''}
+
+        ${optionalString cfg.localBinInPath ''
+          # ~/.local/bin if it exists overrides other bin directories.
```
Comment (SuperSandro2000) [inline nixos/modules/config/shells-environment.nix:211]: that comment is not true. It is not actually checked if the directory exists.

**Review by ryantm — CHANGES_REQUESTED** [review]
With the information I have about this, I'm against inclusion of this. Is this directory a standardized path as the motivation section maybe is hinting at?

**Comment by Kreyren** [comment]
> With the information I have about this, I'm against inclusion of this. Is this directory a standardized path as the motivation section maybe is hinting at? @ryantm

Personal preference plays a role, but the main motivation is scripting since as said `$HOME/bin` is distracting since it shows it in the home dir:

image

If you don't like this integration then i propose better integration of `homeBinInPath` to accept path instead of boolean.

File: nixos/modules/config/shells-environment.nix (hunk 3 above)
Comment (Kreyren) [reply nixos/modules/config/shells-environment.nix:211]: Bug in `homeBinInPath` as well then, will resolve based on convo wit h @ryantm

**Comment by ryantm** [comment]
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

> User-specific executable files may be stored in $HOME/.local/bin. Distributions should ensure this directory shows up in the UNIX $PATH environment variable, at an appropriate place.

**Review by ryantm — APPROVED** [review]

File: nixos/modules/config/shells-environment.nix (hunk 3 above)
Comment (Kreyren) [reply nixos/modules/config/shells-environment.nix]: Resolved in https://github.com/NixOS/nixpkgs/commit/55a211ae3184d105129f00da537c35e14197e300

**Comment by teto** [comment]
thanks for mentioning https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html , I had no idea this was kinda standardized. In which case this PR makes sense.

**Review by teto — APPROVED** [review]

**Pull request merged**